src/models/thread.ts

```diff
@@ -91,85 +99,111 @@ export class Thread extends TypedEventEmitter<ThreadEvent> {
         }
     };
 
+    private get roomState(): RoomState {
+        return this.room.getLiveTimeline().getState(EventTimeline.FORWARDS);
+    }
+
     /**
      * Add an event to the thread and updates
      * the tail/root references if needed
      * Will fire "Thread.update"
      * @param event The event to add
      */
     public async addEvent(event: MatrixEvent, toStartOfTimeline = false): Promise<void> {
-        if (this.timelineSet.findEventById(event.getId())) {
-            return;
+        // When there's no server side support we man
+        if (!this.hasServerSideSupport) {
+            if (this.timelineSet.findEventById(event.getId())) {
+                return;
+            }
+
+            // all the relevant membership info to hydrate events with a sender
+            // is held in the main room timeline
+            // We want to fetch the room state from there and pass it down to this thread
+            // timeline set to let it reconcile an event with its relevant RoomMember
+
+            event.setThread(this);
+            this.timelineSet.addEventToTimeline(
+                event,
+                this.liveTimeline,
+                toStartOfTimeline,
+                false,
+                this.roomState,
+            );
+
+            await this.client.decryptEventIfNeeded(event, {});
         }
 
-        if (!this.root) {
-            if (event.isThreadRelation) {
-                this.root = event.threadRootId;
-            } else {
-                this.root = event.getId();
+        if (this.hasServerSideSupport && this.initialEventsFetched) {
+            if (event.getTs() > this.lastReply().getTs() && !this.findEventById(event.getId())) {
```

Comment:
relying on timestamps between events for ordering is a no-go given different servers will have different clocks

Reply:
What would be the alternative? How can I know whether a message is newly incoming?

Reply:
I'm planning to merge this implementation using the `age`
It has cases in which is could cause the count to drift compared to the value returned by the HS. But this is never going to happen pre-release as we're going to use two homeservers with relatively synced clocks.
I'll create an issue to revisit that before the launch and will see whether we can find a better approach